### ISSUE TITLE
Use Layer Conductance to calculate the attribution score for the attention matrix in BERT

### ISSUE THREAD

**Issue by ZekaiShao25** [pytorch/captum#888]
Hello!
Thank you for this awesome work!
But I was curious about using LG(Layer conductance) to calculate the attribution scores for the attention matrix. 
I don't know much about layer imputation methods, so I checked two papers on LG methods and tried to understand the process of calculating the conductance of neurons. 
From the papers I think that LG uses the gradient of **the final output** of the model to the **input embedding**. So I would like to ask how is the attribution of layer output (hidden_states and attention) done when using LG in BERT to analyze the behavior of the layers? Can you give a clear mathematical formulation?
Of course it's quite possible that I misunderstand the formula in the paper (Eq.7 in [//arxiv.org/pdf/1807.09946.pdf](https://arxiv.org/pdf/1807.09946.pdf))
Thanks for your answer!

**Comment by 99warriors** [comment]
Hi @ZekaiShao25 - thank you for the interest!  At a high level, layer conductance applies neuron conductance to each (output by default) neuron in the specified layer.  Neuron conductance (which gives neuron importance) is an extension of Integrated Gradient (which gives input feature importance, https://arxiv.org/abs/1703.01365), where instead of accumulating (over a path) the change in output `F` when changing input feature `x_i` (i.e. `dF/dx_i`), we accumulate the the change in output `F` when changing input feature `x_i`, due to the change in neuron `y` (i.e. `dF/dy dy/x_i`).  The original Neuron Conductance paper (https://arxiv.org/abs/1805.12233) shows the difference between the 2 methods in equations 1 and 2.  Neuron conductance is well-defined for any neuron, including the attention output.

**Comment by NarineK** [comment]
@ZekaiShao25, to add to what @99warriors mentioned, we have a tutorial in Captum that shows how to apply Layer Conductance on the attention matrices: https://captum.ai/tutorials/Bert_SQUAD_Interpret2

As @99warriors mentioned Layer Conductance is similar to Layer Integrated Gradients. The former applies chain rule w.r.t. the given intermediate layer thus the integration path is slightly different but often Layer Conductance and  Layer Integrated Gradients give similar results.

**Comment by ZekaiShao25** [comment]
Thank you both for your clear reply! I think I have partially understand what @99warriors menthioned, and I have tried the great tutorial @NarineK made. 

But I I have one more question. As mentioned by @99warriors, is the output `F` the attention output of each layer? If so, then the change in the attention output of each hidden layer by changing the input features should be independent of other processing of the final output of the model, i.e. calculating start and end positions. Then why does the tutorial given by @NarineK give different attribution results based on start_logits and end_logits? 

That is, I still don't quite get the relationship among the final attribution goal, layer output and input features.

**Comment by ZekaiShao25** [comment]
Thank you @99warriors !  I think I understand it with your clear comment. Now I close this issue.
